aten/src/ATen/core/op_registration/op_registration.h

```diff
@@ -586,8 +594,196 @@ class CAFFE2_API RegisterOperators final {
+  c10::KernelFunction func_;
+  std::unique_ptr<c10::FunctionSchema> schema_;
+
+  // The "setter" for dispatch_key_
+  template <typename Func>
+  friend CppFunction dispatch(c10::DispatchKey, Func&&);
+
+  // The only class which actually pulls out values from CppFunction (does so
+  // destructively, felt too lazy to write accessors that I don't even
+  // want users to use)
+  friend class Module;
+
+  CppFunction(KernelFunction func, std::unique_ptr<c10::FunctionSchema> schema);
+};
+
+// Create a CppFunction that has "optimized" unboxing wrappers, where
+// the inner unboxed function is inlined into the boxed version of the
+// kernel
+#define TORCH_OPTIMIZED_FN(f) CppFunction::makeOptimized<decltype(f), &f>()
+
+// Create a CppFunction which is associated with a specific dispatch key.
+// CppFunctions that are tagged with a DispatchKey don't get invoked /unless/
+// the dispatcher determines that the DispatchKey is the best choice for
+// a function
+template <typename Func>
+inline CppFunction dispatch(c10::DispatchKey k, Func&& raw_f) {
+  CppFunction f(std::forward<Func>(raw_f));
+  f.dispatch_key_ = k;
+  return f;
+}
+
+// Convenience overload of dispatch which accepts DeviceType
+template <typename Func>
+inline CppFunction dispatch(DeviceType t, Func&& raw_f) {
+  auto deviceTypeToDispatchKey = [](DeviceType t){
+    switch (t) {
+      case DeviceType::CPU:
+        return c10::DispatchKey::CPUTensorId;
+      case DeviceType::CUDA:
+        return c10::DispatchKey::CUDATensorId;
+      case DeviceType::XLA:
+        return c10::DispatchKey::XLATensorId;
+      default:
+        TORCH_CHECK(false,
+          "Device type ", t, " cannot be overloaded at dispatch time, "
+          "please file a bug report explaining what you were trying to do.");
+    }
+  };
+  return dispatch(deviceTypeToDispatchKey(t), std::forward<Func>(raw_f));
+}
+
+// Represents a namespace in which we can define operators.  Conventionally
+// constructed using "torch::import".  This object lets you avoid repeatedly
+// having to specify a namespace, instead you specify it once with:
+//
+//      torch::import("aten")
+//        .def("add", [](const Tensor& a, const Tensor& b) { return a + b; })
+//        .def("mul", torch::dispatch(torch::kCPU, &cpu_add))
+//
+// versus
+//
+//      torch::import()
+//        .def("aten::add", ...)
+//        .def("aten::mul", ...)
+//
+class CAFFE2_API Module final {
+  // TODO: Could store a std::string if you want to support dynamically computed
+  // namespaces; for now don't support
+  const char* ns_;
+
+  // Internal implementation details; right now implement in terms of
+  // RegisterOperators
+  RegisterOperators register_;
+
+  Module(const char* ns);
+
+  // Use these as the constructors
+  friend Module import(const char* ns);
+  friend Module import();
+
+public:
+  Module(const Module&) = delete;
+  Module& operator=(const Module&) = delete;
+
+  Module(Module&&);
```

Comment:
nit: move constructors and assignment operators should be noexcept whenever possible, otherwise STL data types like vector call copy constructors instead even if moving would be possible.